common/src/OpenSpaceNet.cpp

```diff
@@ -102,109 +87,182 @@ OpenSpaceNet::OpenSpaceNet(OpenSpaceNetArgs&& args) :
+    if (labelFilter) {
+        labelFilter->input("predictions") = model->output("predictions");
+        if (nmsNode) {
+            nmsNode->input("predictions") = labelFilter->output("predictions");
+        }
+    } else if (nmsNode) {
+        nmsNode->input("predictions") = model->output("predictions");
+    }
 
-    float confidence = 0;
-    if(args_.action == Action::LANDCOVER && model_->modelOutput() == ModelOutput::BOX) {
-        const auto& blockSize = image_->blockSize();
-        auto windowSize = calcPrimaryWindowSize();
-        if (args_.windowSize.size() < 2) {
-            if(blockSize.width % windowSize.width == 0 &&
-               blockSize.height % windowSize.height == 0) {
-                bbox_ = {cv::Point {0, 0}, image_->size()};
-                concurrent_ = true;
-            }
+    PredictionBoxToPoly::Ptr toPoly;
+    if (!isSegmentation) {
+        toPoly = PredictionBoxToPoly::create("PredictionBoxToPoly");
+        predictionToFeature->input("predictions") = toPoly->output("predictions");
+    }
+
+    if (nmsNode) {
+        if (isSegmentation) {
+            predictionToFeature->input("predictions") = nmsNode->output("predictions");
+        } else {
+            toPoly->input("predictions") = nmsNode->output("predictions");
+        }
+    } else if (labelFilter) {
+        if (isSegmentation) {
+            predictionToFeature->input("predictions") = labelFilter->output("predictions");
+        } else {
+            toPoly->input("predictions") = labelFilter->output("predictions");   
+        }
+    } else {
+        if (isSegmentation) {
+            predictionToFeature->input("predictions") = model->output("predictions");
+        } else {
+            toPoly->input("predictions") = model->output("predictions");      
         }
-    } else if(args_.action == Action::DETECT) {
-        confidence = args_.confidence / 100;
     }
 
-    model_->setConfidence(confidence);
+    if (wfsExtractor) {
+        wfsExtractor->input("features") = predictionToFeature->output("features");
+        featureSink->input("features") = wfsExtractor->output("features");
+    } else {
+        featureSink->input("features") = predictionToFeature->output("features");
+    }
 
-    DG_CHECK(!args_.resampledSize || *args_.resampledSize <= model_->metadata().modelSize().width,
-             "Argument --resample-size (size: %d) does not fit within the model (width: %d).",
-             *args_.resampledSize, model_->metadata().modelSize().width)
+    startProgressDisplay();
+    auto startTime = high_resolution_clock::now();
 
-    if (!args_.resampledSize) {
-        for (auto c : args_.windowSize) {
-            DG_CHECK(c <= model_->metadata().modelSize().width,
-                     "Argument --window-size contains a size that does not fit within the model (width: %d).",
-                      model_->metadata().modelSize().width)
-        }
+    if (!args_.quiet && pd_) {
```

Comment:
You don't need to create a separate thread for each metric, also, detaching threads like this is asking for trouble.

This is what you do:
```cpp
if (!args_.quiet && pd_) {
        Semaphore stop;
        auto allMetrics = async([&stop, &slidingWindow, &featureSink, ...... ] () {
            do {
                if(slidingWindow->metric("processed").changed().tryWait()) {
                    auto processed = slidingWindow->metric("processed").convert<int>();
                    auto requested = slidingWindow->metric("requested").convert<int>();
                    if (processed > 0 && requested > 0) {
                        auto progress = (float) processed / requested;
                        this->pd_->update("Reading", progress);
                    }
                }
                ......

                stop.wait([&slidingWindow, &featureSink, ...... ] () {
                    return featureSink->state() == NodeState::STOPPED ||
                           slidingWindow->metric("processed").changed().check() || 
                           featureSink->metric("processed").changed().check() ||
                           ...... ;
                });

            } while(featureSink->state() != NodeState::STOPPED);
        });

        featureSink->run();
        featureSink->wait();
        stop.notify();
        allMetrics.get();
        pd_->stop();
        ......
```

Reply:
This actually brings up a need to do something better about monitoring `Metric`s. I think the work I'll need to do fix the data race in `Node` will enable something easier to use, but the solution I posted here should suffice for now...